File: csi.m

function csi(data, action, varargin)
% CSI Master Utility for PicoScenes CSI Bundles
%
% Usage:
%   csi(data, 'summary')
%   csi(data, 'plot')
%   csi(data, 'plot', 'frames', 1:200)
%   csi(data, 'export', 'basename', 'csi_export')
%   csi(data, 'subcarriers')
%   csi(data, 'identify_ap_sta')
%
% Where:
%   - data is the PicoScenes bundle (e.g., rx_4_251203_030315)
%   - action is one of:
%       'summary'        → full CSI report
%       'plot'           → visualization toolkit
%       'export'         → export CSI to .mat + .npy
%       'subcarriers'    → frequency mapping for tones
%       'identify_ap_sta'→ heuristic AP/STA identification
%
% NOTE: This utility assumes PicoScenes MATLAB Toolbox structure:
%   data{1} is the main struct with fields:
%       .StandardHeader, .RxSBasic, .CSI, etc.

    if nargin < 2
        error('Usage: csi(data, action, ...)');
    end

    s = data{1};  % main PicoScenes struct

    switch lower(action)
        case 'summary'
            summarize_csi(s);

        case 'plot'
            plot_csi(s, varargin{:});

        case 'export'
            export_csi_to_python(s, varargin{:});

        case 'subcarriers'
            extract_subcarriers(s);

        case 'identify_ap_sta'
            identify_ap_sta(s);

        otherwise
            error('Unknown action: %s', action);
    end
end

% ============================================================
function summarize_csi(s)
% summarize_csi.m → full CSI report (NumTx, NumRx, CBW, sampling rate,
% tones, PHY type, etc.)

    c = s.CSI;

    fprintf('\n==========================================\n');
    fprintf('             CSI SUMMARY REPORT\n');
    fprintf('==========================================\n');

    % Basic CSI fields (if they exist)
    print_field(c, 'DeviceType');
    print_field(c, 'FirmwareVersion');
    print_field(c, 'PacketFormat');
    print_field(c, 'CBW');
    print_field(c, 'CarrierFreq');
    print_field(c, 'CarrierFreq2');
    print_field(c, 'SamplingRate');
    print_field(c, 'SubcarrierBandwidth');
    print_field(c, 'NumTones');
    print_field(c, 'NumTx');
    print_field(c, 'NumRx');
    print_field(c, 'NumESS');
    print_field(c, 'NumCSI');

    % Derived CSI stats
    if isfield(c, 'Mag')
        [nFrames, nTones] = size(c.Mag);
        fprintf('Frames (from Mag): %d\n', nFrames);
        fprintf('Tones  (from Mag): %d\n', nTones);
    end

    % Sampling rate from timestamps (RxSBasic.Timestamp)
    if isfield(s, 'RxSBasic') && isfield(s.RxSBasic, 'Timestamp')
        ts = double(s.RxSBasic.Timestamp); % nanoseconds
        ts_sec = ts / 1e9;
        dt = diff(ts_sec);
        fps = 1 / mean(dt);
        fprintf('Derived CSI sampling rate: %.2f Hz\n', fps);
    else
        fprintf('Derived CSI sampling rate: N/A (no RxSBasic.Timestamp)\n');
    end

    % MAC summary
    try
        [ap_mac, sta_mac] = identify_ap_sta(s);
        fprintf('\nGuessed AP MAC:  %s\n', ap_mac);
        fprintf('Guessed STA MAC: %s\n', sta_mac);
    catch
        fprintf('\nAP/STA identification: not available.\n');
    end

    fprintf('==========================================\n\n');
end

% ============================================================
function plot_csi(s, varargin)
% plot_csi.m → visualization toolkit
%
% Options (name/value pairs):
%   'frames'   → vector of frame indices to use (default: 1:min(200, N))
%   'link'     → [rx, tx] pair index for multi-antenna (ignored here since
%                 Mag/Phase are 2D in many PicoScenes logs)

    p = inputParser;
    addParameter(p, 'frames', [], @(x) isnumeric(x) || islogical(x));
    parse(p, varargin{:});
    frames = p.Results.frames;

    c = s.CSI;

    if ~isfield(c, 'Mag') || ~isfield(c, 'Phase')
        error('CSI structure must contain Mag and Phase fields.');
    end

    Amp = c.Mag;
    Phase = c.Phase;

    [N, K] = size(Amp); % N frames, K subcarriers

    if isempty(frames)
        frames = 1:min(200, N);
    else
        frames = frames(frames >= 1 & frames <= N);
    end

    if isempty(frames)
        error('No valid frames selected for plotting.');
    end

    % ---- Plot amplitude of first selected frame ----
    f1 = frames(1);
    figure;
    plot(Amp(f1, :), 'LineWidth', 1.2);
    xlabel('Subcarrier Index');
    ylabel('Amplitude');
    title(sprintf('CSI Amplitude – Frame %d', f1));
    grid on;

    % ---- Plot phase of first selected frame ----
    figure;
    plot(Phase(f1, :), 'LineWidth', 1.2);
    xlabel('Subcarrier Index');
    ylabel('Phase (radians)');
    title(sprintf('CSI Phase – Frame %d', f1));
    grid on;

    % ---- Amplitude over time heatmap ----
    figure;
    imagesc(Amp(frames, :));
    xlabel('Subcarrier Index');
    ylabel('Frame Index (subset)');
    title('CSI Amplitude Over Time');
    colorbar;

    % ---- Phase over time heatmap ----
    figure;
    imagesc(Phase(frames, :));
    xlabel('Subcarrier Index');
    ylabel('Frame Index (subset)');
    title('CSI Phase Over Time');
    colorbar;
end

% ============================================================
function extract_subcarriers(s)
% extract_subcarriers.m → frequency mapping for tones

    c = s.CSI;

    if ~isfield(c, 'SubcarrierIndex') || ~isfield(c, 'SubcarrierBandwidth')
        error('CSI structure must contain SubcarrierIndex and SubcarrierBandwidth.');
    end

    idx = double(c.SubcarrierIndex);
    delta_f = double(c.SubcarrierBandwidth); % Hz

    freq_offsets_hz = idx * delta_f; % offset from center frequency (Hz)

    fprintf('\n==========================================\n');
    fprintf('        Subcarrier Frequency Mapping\n');
    fprintf('==========================================\n');
    fprintf('Subcarrier spacing: %.2f kHz\n', delta_f / 1e3);
    fprintf('Number of tones:    %d\n\n', numel(idx));

    T = table(idx(:), freq_offsets_hz(:)/1e6, ...
              'VariableNames', {'SubcarrierIndex', 'FreqOffsetMHz'});
    disp(T(1:min(20,height(T)), :)); % show first few

    % Optional plot: mean amplitude vs frequency
    if isfield(c, 'Mag')
        Amp = c.Mag;
        mean_amp = mean(Amp, 1);
        figure;
        plot(freq_offsets_hz/1e6, mean_amp, 'LineWidth', 1.2);
        xlabel('Frequency Offset (MHz)');
        ylabel('Mean Amplitude');
        title('Mean CSI Amplitude vs Frequency Offset');
        grid on;
    end
end

% ============================================================
function export_csi_to_python(s, varargin)
% export_csi_to_python.m → .mat → .npy converter
%
% Saves:
%   basename.mat      → MATLAB file with H, Amp, Phase, subcarriers, ts
%   basename_H.npy    → complex CSI (double, real/imag stacked)
%   basename_Amp.npy  → amplitude (double)
%   basename_Phase.npy→ phase (double)

    p = inputParser;
    addParameter(p, 'basename', 'csi_export', @ischar);
    parse(p, varargin{:});
    basename = p.Results.basename;

    c = s.CSI;

    if ~isfield(c, 'CSI') || ~isfield(c, 'Mag') || ~isfield(c, 'Phase')
        error('CSI structure must contain CSI, Mag, and Phase fields.');
    end

    H = c.CSI;
    Amp = c.Mag;
    Phase = c.Phase;

    % Timestamps (if available)
    ts_sec = [];
    if isfield(s, 'RxSBasic') && isfield(s.RxSBasic, 'Timestamp')
        ts = double(s.RxSBasic.Timestamp); % ns
        ts_sec = ts / 1e9;
    end

    subcarriers = [];
    if isfield(c, 'SubcarrierIndex')
        subcarriers = c.SubcarrierIndex;
    end

    fprintf('\nExporting CSI to %s.mat and .npy files...\n', basename);

    % Save .mat (easy to load in Python via scipy.io.loadmat)
    save([basename '.mat'], 'H', 'Amp', 'Phase', 'subcarriers', 'ts_sec', '-v7.3');

    % Save .npy (simple writer, supports 2D double arrays)
    % For complex H, we save real/imag stacked along 3rd dimension.
    H_reim = cat(3, real(H), imag(H)); % shape: [N, K, 2]

    write_npy([basename '_H.npy'], H_reim);
    write_npy([basename '_Amp.npy'], Amp);
    write_npy([basename '_Phase.npy'], Phase);

    fprintf('Done.\n\n');
    fprintf('Python loading example:\n');
    fprintf('  import numpy as np\n');
    fprintf('  H = np.load("%s_H.npy")   # shape: (N, K, 2)\n', basename);
    fprintf('  Amp = np.load("%s_Amp.npy")\n', basename);
    fprintf('  Phase = np.load("%s_Phase.npy")\n', basename);
    fprintf('  # Or use scipy.io.loadmat("%s.mat")\n\n', basename);
end

% ============================================================
function [ap_mac, sta_mac] = identify_ap_sta(s)
% identify_ap_sta.m → automatically detect AP vs STA roles (heuristic)
%
% Returns:
%   ap_mac  → guessed AP MAC address (string)
%   sta_mac → guessed STA MAC address (string)

    mac2str = @(row) sprintf('%02X:%02X:%02X:%02X:%02X:%02X', row);
    broadcast = 'FF:FF:FF:FF:FF:FF';

    A1 = s.StandardHeader.Addr1; % RA
    A2 = s.StandardHeader.Addr2; % TA
    A3 = s.StandardHeader.Addr3; % often BSSID in infrastructure mode

    n = size(A1,1);

    addr1_str = arrayfun(@(i) mac2str(A1(i,:)), 1:n, 'UniformOutput', false);
    addr2_str = arrayfun(@(i) mac2str(A2(i,:)), 1:n, 'UniformOutput', false);
    addr3_str = arrayfun(@(i) mac2str(A3(i,:)), 1:n, 'UniformOutput', false);

    % Heuristic AP candidate: most frequent MAC in Addr3
    [uniqA3, ~, idx3] = unique(addr3_str);
    countsA3 = accumarray(idx3, 1);
    [~, order3] = sort(countsA3, 'descend');
    ap_mac = uniqA3{order3(1)};

    % Remove broadcast if it somehow appears
    if strcmp(ap_mac, broadcast) && numel(order3) > 1
        ap_mac = uniqA3{order3(2)};
    end

    % STA candidate: most frequent non-AP, non-broadcast MAC appearing in A1/A2
    all_macs = [addr1_str, addr2_str];
    all_macs = all_macs(:);
    all_macs = setdiff(all_macs, {broadcast, ap_mac});

    [uniqAll, ~, idxAll] = unique(all_macs);
    countsAll = accumarray(idxAll, 1);
    [~, orderAll] = sort(countsAll, 'descend');

    sta_mac = uniqAll{orderAll(1)};

    fprintf('\nAP/STA Identification (Heuristic):\n');
    fprintf('  AP  candidate: %s\n', ap_mac);
    fprintf('  STA candidate: %s\n\n', sta_mac);
end

% ============================================================
function print_field(c, fieldname)
% Helper to safely print a field if it exists
    if isfield(c, fieldname)
        val = c.(fieldname);
        if isnumeric(val)
            if isscalar(val)
                fprintf('%s: %g\n', fieldname, val);
            else
                sz = size(val);
                fprintf('%s: numeric [%s]\n', fieldname, num2str(sz));
            end
        elseif ischar(val) || isstring(val)
            fprintf('%s: %s\n', fieldname, string(val));
        else
            sz = size(val);
            fprintf('%s: [%s %s]\n', fieldname, class(val), num2str(sz));
        end
    end
end

% ============================================================
function write_npy(filename, array)
% Minimal NumPy .npy writer for 2D or 3D double arrays (little-endian).
% Supports dtype '<f8' (float64).

    if ~isa(array, 'double')
        array = double(array);
    end

    shape = size(array);
    nd = numel(shape);

    % Build header dict
    shape_str = '(';
    for i = 1:nd
        if i < nd
            shape_str = sprintf('%s%d, ', shape_str, shape(i));
        else
            if nd == 1
                shape_str = sprintf('%s%d,', shape_str, shape(i)); % single dim needs trailing comma
            else
                shape_str = sprintf('%s%d', shape_str, shape(i));
            end
        end
    end
    shape_str = [shape_str, ')'];

    header_dict = sprintf('{' ...
        '''descr'': ''<f8'', ' ...
        '''fortran_order'': False, ' ...
        '''shape'': %s, }', shape_str);

    % NPY header format v1.0
    magic = char([147, 'NUMPY']); %#ok<CHARTEN>
    major = uint8(1);
    minor = uint8(0);

    % Pad header to be aligned to 16 bytes
    header = [header_dict, char(10)]; % newline at end
    header_len = length(header);
    pad_len = 16 - mod(10 + header_len, 16); % 10 = magic(6) + 2 version + 2 header_len
    if pad_len == 16
        pad_len = 0;
    end
    header = [header_dict, repmat(' ', 1, pad_len), char(10)];
    header_len = length(header);

    fid = fopen(filename, 'w');
    if fid == -1
        error('Could not open %s for writing.', filename);
    end

    cleanup = onCleanup(@() fclose(fid));

    fwrite(fid, magic, 'uint8');
    fwrite(fid, major, 'uint8');
    fwrite(fid, minor, 'uint8');

    fwrite(fid, uint16(header_len), 'uint16'); % little-endian by default

    fwrite(fid, header, 'char');

    % Write data in C order
    fwrite(fid, array(:), 'double');
end

File: csi.py
#!/usr/bin/env python3
"""
csi.py  —  Inspect a PicoScenes .csi file (Python version of mac.m)

Requires:
    pip install numpy matplotlib
    PicoScenes Python toolbox installed (module: picoscenes)

Usage:
    python csi.py                 # opens file dialog
    python csi.py path/to/file.csi
"""

import sys
import os

import numpy as np
import matplotlib.pyplot as plt

import tkinter as tk
from tkinter import filedialog, messagebox

from picoscenes import Picoscenes


# ----------------- helpers ----------------- #

def pick_file():
    root = tk.Tk()
    root.withdraw()
    root.update()
    path = filedialog.askopenfilename(
        title="Select PicoScenes .csi file",
        filetypes=[("PicoScenes CSI", "*.csi"), ("All files", "*.*")]
    )
    root.destroy()
    if not path:
        raise SystemExit("No file selected.")
    return path


def mac_to_str(field):
    """
    Convert a 6-byte address (list/ndarray/etc.) to 'AA:BB:CC:DD:EE:FF'.
    Returns None if it cannot be interpreted as a MAC.
    """
    if field is None:
        return None
    try:
        arr = np.array(field, dtype=np.uint8).flatten()
    except Exception:
        return None

    if arr.size < 6:
        return None

    vals = [int(x) & 0xFF for x in arr[:6]]
    return ":".join(f"{b:02X}" for b in vals)


def get_timestamp(rx_basic):
    """
    Return one timestamp in microseconds from RxSBasic dict.
    Tries several key spellings (Timestamp/SystemTime).
    """
    if not isinstance(rx_basic, dict):
        return None

    for key in ("Timestamp", "timestamp", "SystemTime", "systemTime"):
        if key in rx_basic:
            arr = np.array(rx_basic[key]).flatten()
            if arr.size:
                return int(arr[0])
    return None


# ----------------- main analysis ----------------- #

def analyze_csi(path):
    print(f"\nOpening file: {path}")
    print(f"File size: {os.path.getsize(path) / (1024 * 1024):.2f} MB\n")

    ps = Picoscenes(path)
    frames = ps.raw        # list[dict], ALL frames

    # CSI-bearing frames only (for CSI stats)
    csi_frames = [f for f in frames if "CSI" in f]
    num_csi = len(csi_frames)

    if num_csi == 0:
        print("No frames with 'CSI' field found in this file.")
        return

    # ---------- MAC extraction: StandardHeader.Addr1/2/3 over ALL frames ----------
    all_macs = set()
    tx_primary = None
    rx_primary = None

    for f in frames:
        sh = f.get("StandardHeader") or f.get("standardHeader")
        if not isinstance(sh, dict):
            continue

        # Try both capitalizations for the fields
        A1 = sh.get("Addr1") or sh.get("addr1")   # Receiver
        A2 = sh.get("Addr2") or sh.get("addr2")   # Transmitter
        A3 = sh.get("Addr3") or sh.get("addr3")   # BSSID / routing address

        ADDRS = [A1, A2, A3]

        # Convert and filter zeros/broadcast for unique MACs
        for addr in ADDRS:
            mac = mac_to_str(addr)
            if mac is None:
                continue
            if mac in ("00:00:00:00:00:00", "FF:FF:FF:FF:FF:FF"):
                continue
            all_macs.add(mac)

        # Find FIRST non-zero, non-broadcast TX/RX pair
        if tx_primary is None and A1 is not None and A2 is not None:
            rx_mac = mac_to_str(A1)
            tx_mac = mac_to_str(A2)

            if (
                rx_mac not in (None, "00:00:00:00:00:00", "FF:FF:FF:FF:FF:FF")
                and tx_mac not in (None, "00:00:00:00:00:00", "FF:FF:FF:FF:FF:FF")
            ):
                tx_primary, rx_primary = tx_mac, rx_mac

    # ---------- timestamps & CSI rate (only CSI frames) ----------
    timestamps = []
    for f in csi_frames:
        rx_basic = f.get("RxSBasic") or f.get("rxSBasic")
        ts = get_timestamp(rx_basic)
        if ts is not None:
            timestamps.append(ts)

    time_span_sec = None
    csi_rate = None
    if len(timestamps) >= 2:
        t_min = min(timestamps)
        t_max = max(timestamps)
        time_span_sec = (t_max - t_min) / 1e6   # µs → seconds
        if time_span_sec > 0:
            csi_rate = num_csi / time_span_sec

    # ---------- subcarriers ----------
    csi_block = csi_frames[0]["CSI"]
    num_tones = csi_block.get("NumTones") or csi_block.get("numTones")
    sub_idx = np.array(
        csi_block.get("SubcarrierIndex") or csi_block.get("subcarrierIndex") or []
    ).flatten()
    if sub_idx.size:
        num_sub = sub_idx.size
    else:
        num_sub = num_tones

    # ---------- print summary ----------
    print("====== CSI FILE SUMMARY (Python) ======")
    print(f"Total frames parsed      : {len(frames)}")
    print(f"Frames with CSI          : {num_csi}")
    if time_span_sec is not None:
        print(f"Capture time span        : {time_span_sec:.3f} s")
        if csi_rate is not None:
            print(f"Average CSI rate         : {csi_rate:.2f} packets/s")
    else:
        print("Capture time span        : N/A")
    print(f"Subcarriers per CSI frame: {num_sub}")
    print()

    print("==========================================")
    print("        Unique MAC Addresses Found")
    print("==========================================")
    if all_macs:
        for m in sorted(all_macs):
            print(m)
    else:
        print("<none>")

    print("\n==========================================")
    print("        Primary TX / RX MAC Addresses")
    print("==========================================")
    if tx_primary and rx_primary:
        print(f"Transmitter MAC: {tx_primary}")
        print(f"Receiver MAC:   {rx_primary}")
    else:
        print("Could not determine non-broadcast TX/RX pair.")

    print("\nDone.\n")

    # ---------- optional: CSI timeline plot ----------
    if timestamps:
        ts_sorted = sorted(timestamps)
        t0 = ts_sorted[0]
        t_rel = [(t - t0) / 1e6 for t in ts_sorted]   # seconds
        plt.figure(figsize=(8, 4))
        plt.plot(t_rel, range(1, len(t_rel) + 1))
        plt.xlabel("Time since first CSI frame (s)")
        plt.ylabel("Cumulative CSI packets")
        plt.title("CSI Packet Timeline")
        plt.grid(True)
        plt.tight_layout()
        plt.show()


def main():
    if len(sys.argv) > 1:
        path = sys.argv[1]
    else:
        try:
            path = pick_file()
        except SystemExit as e:
            print(e)
            return

    if not os.path.isfile(path):
        print(f"File not found: {path}")
        return

    try:
        analyze_csi(path)
    except Exception as e:
        try:
            root = tk.Tk()
            root.withdraw()
            messagebox.showerror("Error", f"Failed to parse CSI file:\n{e}")
            root.destroy()
        except Exception:
            pass
        print(f"\nERROR while parsing: {e}")


if __name__ == "__main__":
    main()
